File: app/src/main/java/com/elementary/tasks/voice/ConversationAdapter.java
package com.elementary.tasks.voice;

import android.content.Context;
import androidx.databinding.DataBindingUtil;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.elementary.tasks.R;
import com.elementary.tasks.birthdays.BirthdayHolder;
import com.elementary.tasks.birthdays.BirthdayItem;
import com.elementary.tasks.core.utils.Language;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.databinding.AskListItemBinding;
import com.elementary.tasks.databinding.GroupListItemBinding;
import com.elementary.tasks.databinding.ListItemEventsBinding;
import com.elementary.tasks.databinding.NoteListItemBinding;
import com.elementary.tasks.databinding.ReminderListItemBinding;
import com.elementary.tasks.databinding.ShoppingListItemBinding;
import com.elementary.tasks.databinding.ShowReplyLayoutBinding;
import com.elementary.tasks.databinding.SimpleReplyLayoutBinding;
import com.elementary.tasks.databinding.SimpleResponseLayoutBinding;
import com.elementary.tasks.core.data.models.Group;
import com.elementary.tasks.groups.GroupHolder;
import com.elementary.tasks.notes.NoteHolder;
import com.elementary.tasks.notes.NoteItem;
import com.elementary.tasks.reminder.ReminderHolder;
import com.elementary.tasks.reminder.ShoppingHolder;
import com.elementary.tasks.core.data.models.Reminder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import androidx.recyclerview.widget.RecyclerView;

/**
 * Copyright 2017 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class ConversationAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<Reply> mData;
    private Context mContext;
    private ThemeUtil themeUtil;
    private InsertCallback mCallback;

    ConversationAdapter(Context context) {
        this.mContext = context;
        mData = new LinkedList<>();
        themeUtil = ThemeUtil.getInstance(context);
    }

    void setInsertListener(InsertCallback callback) {
        this.mCallback = callback;
    }

    void addReply(Reply reply) {
        if (reply != null) {
            mData.add(0, reply);
            notifyItemInserted(0);
            notifyItemRangeChanged(0, mData.size());
            if (mCallback != null) {
                mCallback.onItemAdded();
            }
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(mContext);
        if (viewType == Reply.REPLY) {
            return new VoiceHolder(SimpleReplyLayoutBinding.inflate(inflater, parent, false).getRoot());
        } else if (viewType == Reply.RESPONSE) {
            return new VoiceResponseHolder(SimpleResponseLayoutBinding.inflate(inflater, parent, false).getRoot());
        } else if (viewType == Reply.REMINDER) {
            return new ReminderHolder(ReminderListItemBinding.inflate(inflater, parent, false).getRoot(), null, themeUtil, false);
        } else if (viewType == Reply.NOTE) {
            return new NoteHolder(NoteListItemBinding.inflate(inflater, parent, false).getRoot(), null);
        } else if (viewType == Reply.GROUP) {
            return new GroupHolder(GroupListItemBinding.inflate(inflater, parent, false).getRoot(), null);
        } else if (viewType == Reply.SHOW_MORE) {
            return new ShowMoreHolder(ShowReplyLayoutBinding.inflate(inflater, parent, false).getRoot());
        } else if (viewType == Reply.BIRTHDAY) {
            return new BirthdayHolder(ListItemEventsBinding.inflate(inflater, parent, false).getRoot(), null, themeUtil);
        } else if (viewType == Reply.SHOPPING) {
            return new ShoppingHolder(ShoppingListItemBinding.inflate(inflater, parent, false).getRoot(), null, themeUtil);
        } else if (viewType == Reply.ASK) {
            return new AskHolder(AskListItemBinding.inflate(inflater, parent, false).getRoot());
        }
        return null;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof VoiceHolder) {
            ((VoiceHolder) holder).binding.replyText.setText((String) mData.get(position).getObject());
        } else if (holder instanceof VoiceResponseHolder) {
            ((VoiceResponseHolder) holder).binding.replyText.setText((String) mData.get(position).getObject());
        } else if (holder instanceof ReminderHolder) {
            ((ReminderHolder) holder).setData((Reminder) mData.get(position).getObject());
        } else if (holder instanceof NoteHolder) {
            ((NoteHolder) holder).setData((NoteItem) mData.get(position).getObject());
        } else if (holder instanceof GroupHolder) {
            ((GroupHolder) holder).setData((Group) mData.get(position).getObject());
        } else if (holder instanceof BirthdayHolder) {
            ((BirthdayHolder) holder).setData((BirthdayItem) mData.get(position).getObject());
            ((BirthdayHolder) holder).setColor(themeUtil.getColor(themeUtil.colorBirthdayCalendar()));
        } else if (holder instanceof ShoppingHolder) {
            ((ShoppingHolder) holder).setData((Reminder) mData.get(position).getObject());
        } else if (holder instanceof AskHolder) {
            ((AskHolder) holder).setAskAction((AskAction) mData.get(position).getObject());
        }
    }

    @Override
    public int getItemViewType(int position) {
        return mData.get(position).getViewType();
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public void removeAsk() {
        for (int i = 1; i < mData.size(); i++) {
            Reply reply = mData.get(i);
            if (reply != null && reply.getViewType() == Reply.ASK) {
                mData.remove(i);
                notifyItemRemoved(i);
                notifyItemRangeChanged(0, mData.size());
                break;
            }
        }
    }

    private class AskHolder extends RecyclerView.ViewHolder {

        private AskListItemBinding binding;
        private AskAction askAction;

        AskHolder(View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
            binding.replyYes.setOnClickListener(v -> {
                removeFirst();
                if (askAction != null) {
                    askAction.onYes();
                }
            });
            binding.replyNo.setOnClickListener(v -> {
                removeFirst();
                if (askAction != null) {
                    askAction.onNo();
                }
            });
            binding.replyNo.setBackgroundResource(themeUtil.getRectangle());
            binding.replyYes.setBackgroundResource(themeUtil.getRectangle());
            binding.replyNo.setText(Language.getLocalized(mContext, R.string.no));
            binding.replyYes.setText(Language.getLocalized(mContext, R.string.yes));
        }

        void setAskAction(AskAction askAction) {
            this.askAction = askAction;
        }
    }

    private void removeFirst() {
        mData.remove(0);
        notifyItemRemoved(0);
        notifyItemRangeChanged(0, mData.size());
    }

    private class VoiceHolder extends RecyclerView.ViewHolder {

        SimpleReplyLayoutBinding binding;

        VoiceHolder(View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
        }
    }

    private class VoiceResponseHolder extends RecyclerView.ViewHolder {

        SimpleResponseLayoutBinding binding;

        VoiceResponseHolder(View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
        }
    }

    private class ShowMoreHolder extends RecyclerView.ViewHolder {

        private ShowReplyLayoutBinding binding;

        ShowMoreHolder(View itemView) {
            super(itemView);
            binding = DataBindingUtil.bind(itemView);
            binding.replyText.setOnClickListener(view -> addMoreItemsToList(getAdapterPosition()));
        }
    }

    private void addMoreItemsToList(int position) {
        Reply reply = mData.get(position);
        Container container = (Container) reply.getObject();
        if (container.getType() instanceof Group) {
            mData.remove(position);
            notifyItemRemoved(position);
            //noinspection unchecked
            for (Group item : ((Container<Group>) container).getList()) {
                mData.add(0, new Reply(Reply.GROUP, item));
                notifyItemInserted(0);
            }
            notifyItemRangeChanged(0, mData.size());
            if (mCallback != null) mCallback.onItemAdded();
        } else if (container.getType() instanceof NoteItem) {
            mData.remove(position);
            notifyItemRemoved(position);
            //noinspection unchecked
            for (NoteItem item : ((Container<NoteItem>) container).getList()) {
                mData.add(0, new Reply(Reply.NOTE, item));
                notifyItemInserted(0);
            }
            notifyItemRangeChanged(0, mData.size());
            if (mCallback != null) mCallback.onItemAdded();
        } else if (container.getType() instanceof Reminder) {
            mData.remove(position);
            notifyItemRemoved(position);
            addRemindersToList(container);
            notifyItemRangeChanged(0, mData.size());
            if (mCallback != null) mCallback.onItemAdded();
        } else if (container.getType() instanceof BirthdayItem) {
            mData.remove(position);
            notifyItemRemoved(position);
            //noinspection unchecked
            List<BirthdayItem> reversed = new ArrayList<>(((Container<BirthdayItem>) container).getList());
            Collections.reverse(reversed);
            for (BirthdayItem item : reversed) {
                mData.add(0, new Reply(Reply.BIRTHDAY, item));
                notifyItemInserted(0);
            }
            notifyItemRangeChanged(0, mData.size());
            if (mCallback != null) mCallback.onItemAdded();
        }
    }

    private void addRemindersToList(Container container) {
        //noinspection unchecked
        List<Reminder> reversed = new ArrayList<>(((Container<Reminder>) container).getList());
        Collections.reverse(reversed);
        for (Reminder item : reversed) {
            if (item.getViewType() == Reminder.REMINDER) {
                mData.add(0, new Reply(Reply.REMINDER, item));
            } else {
                mData.add(0, new Reply(Reply.SHOPPING, item));
            }
            notifyItemInserted(0);
        }
    }

    interface InsertCallback {
        void onItemAdded();
    }
}

File: app/src/main/java/com/elementary/tasks/voice/ConversationAdapter.kt
package com.elementary.tasks.voice

import android.os.Handler
import android.os.Looper
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.elementary.tasks.R
import com.elementary.tasks.birthdays.list.BirthdayHolder
import com.elementary.tasks.core.arch.CurrentStateHolder
import com.elementary.tasks.core.binding.HolderBinding
import com.elementary.tasks.core.data.models.NoteWithImages
import com.elementary.tasks.core.data.models.ReminderGroup
import com.elementary.tasks.core.data.ui.UiBirthdayList
import com.elementary.tasks.core.data.ui.UiReminderListActive
import com.elementary.tasks.core.data.ui.UiReminderListActiveShop
import com.elementary.tasks.core.utils.inflater
import com.elementary.tasks.databinding.ListItemAskBinding
import com.elementary.tasks.databinding.ListItemShowReplyBinding
import com.elementary.tasks.databinding.ListItemSimpleReplyBinding
import com.elementary.tasks.databinding.ListItemSimpleResponseBinding
import com.elementary.tasks.groups.list.GroupHolder
import com.elementary.tasks.notes.list.NoteViewHolder
import com.elementary.tasks.notes.preview.ImagesSingleton
import com.elementary.tasks.reminder.lists.adapter.ReminderViewHolder
import com.elementary.tasks.reminder.lists.adapter.ShoppingViewHolder
import timber.log.Timber

class ConversationAdapter(
  private val currentStateHolder: CurrentStateHolder,
  private val imagesSingleton: ImagesSingleton
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

  private val data = mutableListOf<Reply>()
  var showMore: ((Int) -> Unit)? = null
  private val handler = Handler(Looper.getMainLooper())

  fun submitList(list: List<Reply>?) {
    Timber.d("submitList: $list")
    if (list == null) {
      this.data.clear()
      notifyDataSetChanged()
      return
    }
    val oldSize = this.data.size
    val diffResult = DiffUtil.calculateDiff(ReplyDiffCallback(this.data, list))
    this.data.clear()
    this.data.addAll(list)
    diffResult.dispatchUpdatesTo(this)
    handler.postDelayed({
      if (!list.isNullOrEmpty()) {
        runCatching {
          if (oldSize == list.size) {
            notifyItemChanged(0)
          } else {
            notifyDataSetChanged()
          }
        }
      }
    }, 250)
  }

  fun getItem(position: Int): Reply? =
    if (position != -1 && position < data.size) data[position] else null

  override fun getItemCount(): Int {
    return data.size
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
    return when (viewType) {
      Reply.REPLY -> VoiceHolder(parent)
      Reply.RESPONSE -> VoiceResponseHolder(parent)
      Reply.REMINDER -> ReminderViewHolder(parent, editable = false, showMore = false)
      Reply.NOTE -> NoteViewHolder(parent, currentStateHolder, imagesSingleton, null)
      Reply.GROUP -> GroupHolder(parent, null)
      Reply.SHOW_MORE -> ShowMoreHolder(parent)
      Reply.BIRTHDAY -> BirthdayHolder(parent, showMore = false)
      Reply.SHOPPING -> ShoppingViewHolder(
        parent,
        editable = false,
        showMore = false,
        isDark = currentStateHolder.theme.isDark
      )

      else -> AskHolder(parent)
    }
  }

  override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
    val content = getItem(position)?.content
    when {
      holder is VoiceHolder -> holder.bind(content as String)
      holder is VoiceResponseHolder -> holder.bind(content as String)
      holder is ReminderViewHolder -> holder.setData(content as UiReminderListActive)
      holder is NoteViewHolder && content is NoteWithImages -> holder.setData(content)
      holder is GroupHolder -> holder.setData(content as ReminderGroup)
      holder is BirthdayHolder -> holder.setData(content as UiBirthdayList)
      holder is ShoppingViewHolder -> holder.setData(content as UiReminderListActiveShop)
      holder is AskHolder -> holder.setAskAction(content as AskAction)
    }
  }

  override fun getItemViewType(position: Int): Int {
    return getItem(position)?.viewType ?: 0
  }

  private inner class AskHolder(
    parent: ViewGroup
  ) : HolderBinding<ListItemAskBinding>(
    ListItemAskBinding.inflate(parent.inflater(), parent, false)
  ) {

    private var askAction: AskAction? = null

    init {
      binding.replyYes.setOnClickListener {
        askAction?.onYes()
      }
      binding.replyNo.setOnClickListener {
        askAction?.onNo()
      }
      binding.replyNo.text = currentStateHolder.language.getLocalized(itemView.context, R.string.no)
      binding.replyYes.text =
        currentStateHolder.language.getLocalized(itemView.context, R.string.yes)
    }

    fun setAskAction(askAction: AskAction) {
      this.askAction = askAction
    }
  }

  private inner class VoiceHolder(
    parent: ViewGroup
  ) : HolderBinding<ListItemSimpleReplyBinding>(
    ListItemSimpleReplyBinding.inflate(parent.inflater(), parent, false)
  ) {
    fun bind(text: String) {
      binding.replyTextSimple.text = text
    }
  }

  private inner class VoiceResponseHolder(
    parent: ViewGroup
  ) : HolderBinding<ListItemSimpleResponseBinding>(
    ListItemSimpleResponseBinding.inflate(parent.inflater(), parent, false)
  ) {
    fun bind(text: String) {
      binding.replyTextResponse.text = text
    }
  }

  private inner class ShowMoreHolder(
    parent: ViewGroup
  ) : HolderBinding<ListItemShowReplyBinding>(
    ListItemShowReplyBinding.inflate(parent.inflater(), parent, false)
  ) {
    init {
      binding.replyText.setOnClickListener {
        showMore?.invoke(bindingAdapterPosition)
        showMore = null
      }
    }
  }
}
